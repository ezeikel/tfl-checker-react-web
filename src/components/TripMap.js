import React from "react";
import PropTypes from "prop-types";
import styled from "styled-components";
import GoogleMapReact from "google-map-react";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";

const Wrapper = styled.section`
  height: 100%;
  width: 100%;
  border-radius: 4px;
  overflow: hidden;
`;

const IconWrapper = styled.span`
  position: absolute;
  transform: translate(-50%, -100%);
  z-index: 10;
`;

const defaultProps = {
  zoom: 11,
};

const handleGoogleMapApi = (google, path) => {
  const bounds = new google.maps.LatLngBounds();

  const tripPath = new google.maps.Polyline({
    path,
    geodesic: true,
    strokeColor: "#00BCD4",
    strokeOpacity: 1,
    strokeWeight: 5,
  });

  tripPath.setMap(google.map);

  bounds.extend(path[0]);
  bounds.extend(path[path.length - 1]);

  google.map.fitBounds(bounds);
};

const TripMap = ({ center, path }) => (
  <Wrapper>
    <GoogleMapReact
      defaultCenter={center}
      defaultZoom={defaultProps.zoom}
      yesIWantToUseGoogleMapApiInternals
      onGoogleApiLoaded={google => handleGoogleMapApi(google, path)}
    >
      <IconWrapper lat={path[0].lat} lng={path[0].lng}>
        <FontAwesomeIcon
          icon={["fas", "map-marker-alt"]}
          color="#31CC71"
          size="3x"
        />
      </IconWrapper>
      <IconWrapper
        lat={path[path.length - 1].lat}
        lng={path[path.length - 1].lng}
      >
        <FontAwesomeIcon
          icon={["fas", "map-marker-alt"]}
          color="#00BCD4"
          size="3x"
        />
      </IconWrapper>
    </GoogleMapReact>
  </Wrapper>
);

TripMap.propTypes = {
  center: PropTypes.shape({
    lat: PropTypes.number,
    lng: PropTypes.number,
  }).isRequired,
  path: PropTypes.arrayOf(
    PropTypes.shape({
      lat: PropTypes.number,
      lng: PropTypes.number,
    }),
  ).isRequired,
};

export default TripMap;
