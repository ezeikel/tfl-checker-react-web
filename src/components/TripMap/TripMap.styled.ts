import styled from "styled-components";

export const Wrapper = styled.section`
  height: 100%;
  width: 100%;
  border-radius: 4px;
  overflow: hidden;
`;

export const IconWrapper = styled.span`
  position: absolute;
  transform: translate(-50%, -100%);
  z-index: 10;
`;
