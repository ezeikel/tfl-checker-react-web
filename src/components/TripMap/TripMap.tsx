import GoogleMapReact from "google-map-react";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { Wrapper, IconWrapper } from "./TripMap.styled";

type TripMapProps = {
  center: {
    lat: number;
    lng: number;
  };
  path: {
    lat: number;
    lng: number;
  };
};

const defaultProps = {
  zoom: 11,
};

const handleGoogleMapApi = (google, path) => {
  const bounds = new google.maps.LatLngBounds();

  const tripPath = new google.maps.Polyline({
    path,
    geodesic: true,
    strokeColor: "#00BCD4",
    strokeOpacity: 1,
    strokeWeight: 5,
  });

  tripPath.setMap(google.map);

  bounds.extend(path[0]);
  bounds.extend(path[path.length - 1]);

  google.map.fitBounds(bounds);
};

const TripMap = ({ center, path }: TripMapProps) => (
  <Wrapper>
    <GoogleMapReact
      defaultCenter={center}
      defaultZoom={defaultProps.zoom}
      yesIWantToUseGoogleMapApiInternals
      onGoogleApiLoaded={(google) => handleGoogleMapApi(google, path)}
    >
      <IconWrapper lat={path[0].lat} lng={path[0].lng}>
        <FontAwesomeIcon
          icon={["fas", "map-marker-alt"]}
          color="#31CC71"
          size="3x"
        />
      </IconWrapper>
      <IconWrapper
        lat={path[path.length - 1].lat}
        lng={path[path.length - 1].lng}
      >
        <FontAwesomeIcon
          icon={["fas", "map-marker-alt"]}
          color="#00BCD4"
          size="3x"
        />
      </IconWrapper>
    </GoogleMapReact>
  </Wrapper>
);

export default TripMap;
