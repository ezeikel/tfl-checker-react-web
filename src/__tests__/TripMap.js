import React from 'react';
import { render } from '@testing-library/react';
import TripMap from '../components/TripMap';

let mockProps;

beforeEach(() => {
  mockProps = {
    center: {
      lat: 0,
      lng: 0
    },
    path: [
      {
        lat: 0,
        lng: 0
      }
    ]
  };
});

test('renders without crashing', () => {
  const { container } = render(
    <TripMap {...mockProps} /> // eslint-disable-line react/jsx-props-no-spreading
  );
  expect(container).toMatchSnapshot();
});
