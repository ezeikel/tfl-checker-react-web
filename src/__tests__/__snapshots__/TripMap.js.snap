// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders without crashing 1`] = `
<div>
  <section
    class="sc-AxjAm lguFjZ"
  >
    <div
      style="width: 100%; height: 100%; margin: 0px; padding: 0px; position: relative;"
    >
      <div
        style="width: 100%; height: 100%; left: 0px; top: 0px; margin: 0px; padding: 0px; position: absolute;"
      />
      <div
        style="width: 50%; height: 50%; left: 50%; top: 50%; margin: 0px; padding: 0px; position: absolute;"
      >
        <div
          style="width: 100%; height: 100%; left: 0px; top: 0px; margin: 0px; padding: 0px; position: absolute;"
        >
          <div
            style="width: 0px; height: 0px; left: 0px; top: 0px; background-color: transparent; position: absolute;"
          >
            <span
              class="sc-AxirZ fjeuzW"
            >
              <svg
                class="svg-inline--fa fas-map-marker-alt fa-3x"
                color="#31CC71"
              />
            </span>
          </div>
          <div
            style="width: 0px; height: 0px; left: 0px; top: 0px; background-color: transparent; position: absolute;"
          >
            <span
              class="sc-AxirZ fjeuzW"
            >
              <svg
                class="svg-inline--fa fas-map-marker-alt fa-3x"
                color="#00BCD4"
              />
            </span>
          </div>
        </div>
      </div>
    </div>
  </section>
</div>
`;
